# New skills{-}
================================================================================

>__About the exercise__

<br>
In this Rmarkdown file I will explain the new skills I learned (this is mentioned in the previous chapter). The goal is to learn a new skill in 4 days of self study, I decided to divide my 4 days in to 2x2 days to learn the following skills:

- Microbiota analysis         ([link to the workflow](https://rstudio-pubs-static.s3.amazonaws.com/268156_d3ea37937f4f4469839ab6fa2c483842.html#other_visualizations))
- Shiny                      ([link to the workflow](https://mastering-shiny.org/basic-app.html))
<br>

I came up with an assignment to test my newly learned skills, the assignment is to analyse a dataset (this dataset is from my last project) and apply some plots on the samples. After that I make the plots visible in a Shiny app. In the app I will ask for 3 different inputs that the user can adjust. 
<br>

***

>__Microbiota analysis__


The goal of this workflow is to learn basic analyses of microbiota data to determine if and how communities differ by variables of interest. This pipeline can be used for any microbiota dataset that has been clustered into operational taxonomic units: OTU's. All the resultes, code and pipeline are documented in the file: code/Microbiota_analysis.R. In this workflow I will learn about alpha and beta-diversity between and within groups. After that there are different kind of graphs introduced (Abundance plots, heat maps, etc.).
<br>

***

>__Shiny__


The second skill I will learn about is Shiny. Shiny is a package that can be installed in Rstudio, it makes it easy to build interactive web apps straight from Rstudio. To show the plots you created in a project it is important that the figures are interpreted clearly and comprehensibly. Shiny gives you the opportunity to do this in an app with interactive plots. My goal of this workflow is to learn the basics of Shiny and make a small app that shows you the results of my last project in an interactive way. Below I will give you a quick introduction of my last project.
<br>

***

>__Summary of my last project__

The data used in this workflow is from my last project. Here is a short summary of the project:

_The Waternet project is a project commissioned by Waternet. It's aim is to get insight in the bacterial composition of Waternet's water, before and after filtering. The project was initially handled by a group op microbiology students. Both mock solutions and samples were used for determination, the latter as a control on the determination process. After concluding they had insufficient resources to properly determine the composition of the samples, the samples were sent to a company that performed whole genome sequencing on them. As a continuation of this project, a group of data science students (us) were assigned the task of coming up with a way of determining the bacterial composition of both samples, based on the WGS data._

__For more information about the project visit our [GitHub](https://github.com/ProjecticumDataScience/Project-1-Waternet) page.__
<br><br>
The samples I am going to analyse are the first 10 000 reads of the pre and post sample (this will be ajustable via the interface). I am gonna visualize them in different ways so that the data will be understandable to interpret. If you are working with Shiny it's a good idea to start with a sketch. In the image below you can see mine.
<br>

```{r figure_shiny, out.width= "400px", out.height= "600px", echo=FALSE}
knitr::include_graphics(here::here("images", "shiny_sketch.jpg"))
```

***

>__Shiny app__

I decided to put both samples (pre and post) next to eachother in the app, so that you will get a quick overview what is different in each sample. I also made use of tabs so that the table and plots are not on the same page. This makes it easier to view the data. To start the app you need to open the file _app/waternet_app.R_ and press on the button: _Run App_. The app is best viewed in full screen.

